### PR TITLE
Loki: Use better query validation before requesting stats

### LINKED ISSUE #70774: Loki: Display query stats for queries containing range variables
Currently Loki queries with range variables like `$__interval` or `$__range` are considered "invalid" queries, thus the query stats are not requested. 
I think we should replace those variables, or create another validation not considering range values, to also display query stats for a query like:

```
sum(rate({place="luna"} [$__interval]))
```

### PR DESCRIPTION
**Pull request by itsgareth**
**What is this feature?**
Uses a better validation function before requesting stats
- before stats where not requested if the query had any variables
- now stats are requested for queries with variables

Additional change:
- Trim queries before comparing, to prevent spamming stats endpoint.
- Added missing coverage for getQueryStats.

**Which issue(s) does this PR fix?**:
Fixes #70774

### CODE REVIEW

File: public/app/plugins/datasource/loki/datasource.ts [hunk 1]
```diff
@@ -451,7 +455,8 @@ export class LokiDatasource
 
   async getQueryStats(query: string): Promise<QueryStats | undefined> {
     // if query is invalid, clear stats, and don't request
-    if (!isValidQuery(query)) {
+    const errors = validateQuery(query, this.interpolateString(query, placeHolderScopedVars), []);
+    if (typeof errors === 'object') {
       return undefined;
     }
```
Comment (matyax) [inline public/app/plugins/datasource/loki/datasource.ts:461]: I regret the way I designed this function, it's really awkward to consume. 

```suggestion
    const errors = validateQuery(query, this.interpolateString(query, placeHolderScopedVars), []);
    if (errors !== false) {
      return undefined;
    }
```

You can also use `Array.isArray()`.

But even better: would you create a new function inside `validation.ts`, with the same signature of `validateQuery`, but it returns a boolean? E.g.

```javascript
function isQueryValid(...): boolean {
  return validateQuery(...) === false ? true : false;
}
```

File: public/app/plugins/datasource/loki/datasource.ts [hunk 2]
```diff
@@ -451,7 +455,8 @@ export class LokiDatasource
 
   async getQueryStats(query: string): Promise<QueryStats | undefined> {
```
Comment (matyax) [inline public/app/plugins/datasource/loki/datasource.ts:456]: https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/loki/datasource.test.ts#L1200-L1207

+ 

```javascript
it("validates queries with variables", () => {});
it("does not call stats if the query is invalid", () => {});
```

**Review by matyax — COMMENTED** [review]
Left some suggestions. Also remember to rename `isValidQuery` from `queryUtils` to `isQueryWithError`.

File: public/app/plugins/datasource/loki/datasource.ts (hunk 1 above)
Comment (gwdawson) [reply public/app/plugins/datasource/loki/datasource.ts:461]: @matyax looking at `validateQuery()`, it doesn't seem like it is used that many places, would it be better to just change the response of this function directly.

Comment (matyax) [reply public/app/plugins/datasource/loki/datasource.ts:461]: Good question. I think `validateQuery` is correctly serving its purpose: returning an array of errors. The only change that I would make is to always return an array, empty or with errors.

Then, I think we're missing a function that we (or other developers) can call ot get a boolean indicating if the query is valid or not.

File: public/app/plugins/datasource/loki/datasource.test.ts [hunk 3]
```diff
@@ -1198,11 +1198,43 @@ describe('LokiDatasource', () => {
...
+    });
+
+    it('does not call stats if the query is invalid', async () => {
+      const result = await ds.getQueryStats('rate({label="value"}');
+
+      expect(ds.statsMetadataRequest).not.toHaveBeenCalled();
+      expect(result).toBe(undefined);
+    });
+
+    it('combines the stats of each label matcher', async () => {
```
Comment (matyax) [inline public/app/plugins/datasource/loki/datasource.test.ts:1233]: @gwdawson Just saw that there was missing coverage for this part of the code:

https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/loki/datasource.ts#L459-L463

Is that correct? Should you accumulate the stats of `foo => bar` and `test => test` in a query like `count_over_time({foo="bar"}[1m]) + count_over_time({test="test"}[1m])`?

**Comment by matyax** [comment]
I'm going to skip reviewing this PR because I've commited to this branch.

File: public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/validation.ts [hunk 4]
```diff
@@ -16,6 +16,23 @@ interface ParseError {
...
 
+/**
+ * Simplified version of `validateQuery` to determine if a given query is valid or not.
+ * Grafana variables are errors to the parser, so it is ideal to call this function with the
+ * interpolated query.
+ * @alpha
+ */
+export function isValidQuery(interpolatedQuery: string): boolean {
+  return validateQuery(interpolatedQuery, interpolatedQuery, interpolatedQuery.split('\\n')) === false ? true : false;
+}
```
Comment (svennergr) [inline public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/validation.ts:27]: I might oversee something, but what's the better way to check if a query is valid now? Do we need both cases?

```ts
isValidQuery(this.interpolateString(query, placeHolderScopedVars))
```

or 

```ts
!isQueryWithError(this.interpolateString(query, placeHolderScopedVars))
```

Comment (matyax) [reply public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/validation.ts:27]: `isValidQuery` is the one. `isQueryWithError` looks for error nodes. I actually thought of naming it `isQueryWithErrorNode` to be more explicit.

Comment (matyax) [reply public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/validation.ts:27]: We should probably also use it in the Datasource too, no?

```
async getDataSamples(query: LokiQuery): Promise<DataFrame[]> {
    // Currently works only for logs sample
    if (isQueryWithError(query.expr) || !isLogsQuery(query.expr)) {
      return [];
    }
    // ...
```

Comment (matyax) [reply public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/validation.ts:27]: I mean, if the query has a variable we will not request samples.

Comment (svennergr) [reply public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/validation.ts:27]: > isValidQuery is the one. isQueryWithError looks for error nodes.

But what's the logical difference in running both on an interpolated query expression? Both should return the same, right? Well, `isValidQuery(interpolatedExpr) !== isQueryWithError(interpolatedExpr)` would be always true.

**Review by svennergr — APPROVED** [review]
LGTM!

**Pull request merged**